### PR TITLE
Remove case sensitivity of Etag header for use_etag

### LINKED ISSUE #61440: [BUG] use_etag expects case-sensitive header Etag
**Description**
Logic omission found in newly merged code #61391 

> Does this account for header case? The standard is `ETag` but I'd expect it to be normalised to `etag`, not `Etag`.

_Originally posted by @OrangeDog in https://github.com/saltstack/salt/issues/61391#issuecomment-1009434165_

**Expected behavior**
apache2 and `http.server` use `Etag`, while nginx uses `ETag`. Current logic would only work on the former. We shouldn't care either way.

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
See issue.

### What issues does this PR fix or reference?
Fixes: #61440

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by nicholasmhughes** [comment]
re-run full all

File: salt/fileclient.py [hunk 1]
```diff
@@ -696,9 +696,19 @@ def on_header(hdr):
                     # Try to find out what content type encoding is used if
                     # this is a text file
                     write_body[1].parse_line(hdr)  # pylint: disable=no-member
-                    if use_etag and "Etag" in write_body[1]:
+                    # Case insensitive Etag header checking below. Don't break case
+                    # insensitivity unless you really want to mess with people's heads
+                    # in the tests. Note: http.server and apache2 use "Etag" and nginx
+                    # uses "ETag" as the header key. Yay standards!
```
Comment (OrangeDog) [inline salt/fileclient.py:702]: The standard says that all HTTP header names are case-insensitive, so it doesn't matter what casing they use.

**Comment by nicholasmhughes** [comment]
re-run full all

File: tests/pytests/functional/states/test_file.py [hunk 2]
```diff
@@ -53,7 +54,13 @@ def do_GET(self):
 
         # Return the Etag header if we have the checksum
         if checksum:
-            self.send_header("Etag", checksum)
+            # IMPORTANT: This introduces randomness into the tests. The Etag header key
+            # will be converted to lowercase in the code... but if someone breaks that,
+            # it'll rear it's head here as random failures that are hard to reproduce.
+            # Any alternatives seem overly complex. So... don't break the case insensitivity
+            # in the code.
+            possible_etags = ["Etag", "ETag"]
```
Comment (waynew) [inline tests/pytests/functional/states/test_file.py:62]: (Hilarious) **suggestion(non-blocking, entirely for the lulz)**: eTaG :trollface:

**Review by waynew — APPROVED** [review]
:+1:

**Comment by waynew** [comment]
re-run all

File: tests/pytests/functional/states/test_file.py (hunk 2 above)
Comment (OrangeDog) [reply tests/pytests/functional/states/test_file.py:62]: More seriously, `etag` is certainly likely in the wild.

Also, can we cut the comments insinuating that this is some sort of workaround, when it's what every system that receives HTTP responses must do by default?

I am surprised that the HTTP client isn't normalising header names itself already, or providing some sort of case-insensitive map. That's what e.g. `requests` does.

**Comment by damon-atkins** [comment]
Hi @nicholasmhughes thanks for doing this. A better method would be to write the whole header to a different file instead of using a file specific for etag. There are many ways to determine if a file should be cached from a web server. Keeping the whole header in a filename.meta file allows not only etag to be used but other techniques, like if-not-modified-since using the date from save header  which is in the same format as what the web server provided (instead of try to create it). Instead of use_etag just have http_cache. e.g.

Here is some sample loading back the data into a case insensitive structure.
meta_file = '{0}.meta'.format(dest_file)
with salt.utils.files.fopen(meta_file, 'r', encoding='utf-8') as meta_fh:
   cache_headers = requests.structures.CaseInsensitiveDict(data=salt.utils.json.load(meta_fh))
Not sure why salt does not use the request library.

Who are you handling failures/partial downloads?

**Review by twangboy — APPROVED** [review]

**Comment by nicholasmhughes** [comment]
@damon-atkins , totally understand the broader metadata perspective. However, I'm not really in a spot to address that at this time.

Regarding handling failures/partial downloads, the code additions for `use_etag` aren't in the critical path for content download handling itself... just the decision on whether or not to download. If an etag file isn't there, is corrupted, or otherwise "partial", then it'll be treated as not matching and re-download the content.

**Pull request merged**